### ISSUE TITLE
Overriding #create.

### ISSUE THREAD

**Issue by beauby** [hanami/model#349]
The way `Hanami::Repository` defines the standard query methods on subclasses seems not to be compatible with overriding those (`super` fails).

```ruby
class UserRepository < Hanami::Repository
  def create(user)
    user[:encrypted_password] =
      BCrypt::Password.create(user.delete(:password))
    super(user)
  end
end
```

```
     # NoMethodError:
     #   super: no superclass method `create' for #<UserRepository:0x007f9196193f78>
```

**Comment by cllns** [comment]
Hm, this is very similar to a `before_create` callback. 

Instead, how do you feel about writing an `Interactor` that creates a User and encrypts their password?

**Comment by beauby** [comment]
@cllns I can't say that I've seen any mention of `Interactor`s in the guides – do you have a link?

Regardless, I'm curious why the default commands are implemented the way they are (i.e. by dynamically `prepend`ing a module during class inheritance, rather than simply defining them in the parent class – my guess is that it either has to do with the fact that `Hanami::Repository` was initially intended to be a mixin rather than a parent class, or with something ROM does related to those default commands).

The current set up leads to the following inheritance chain, for which I see no good reason.
```
> UserRepository.ancestors
 => [Hanami::Repository::Commands, UserRepository, Hanami::Utils::ClassAttribute, Hanami::Repository, ROM::Repository::Root, ROM::Repository, Object, Hanami::Commands::Console::CodeReloading, JSON::Ext::Generator::GeneratorMethods::Object, Kernel, BasicObject] 
```

**Comment by cllns** [comment]
There's [no guide for Interactors yet](https://github.com/hanami/hanami.github.io/issues/215) but you can see the docs here: http://www.rubydoc.info/gems/hanami-utils/Hanami/Interactor

You don't have to use `Hanami::Interactor`, and 'service' object will do.

**Comment by beauby** [comment]
@cllns Thanks for the reading material :+1:.

After a bit of investigation, the commands are created dynamically by the call to `commands`, so the definition of the overriding of the methods that Hanami does must happen after that. I understand why the `prepend` would prevent overriding the overrides that `Hanami::Repository` does, but it's not clear to me why in the above example `super` does not refer to the ROM::Repository dynamically created method `create`.

**Comment by solnic** [comment]
I wouldn't recommend overriding core methods like `create` and introduce your own APIs that use `create`. Regardless though, this is a violation of POLS so we gotta fix that in rom-repository (I was actually aware of that, just didn't have time to make it better).

**Comment by mereghost** [comment]
+1 for using `Hanami::Interactors`. =D

**Comment by jodosha** [comment]
Also, `Hanami::Repository` adds another `#create` on top of `ROM::Repository`, by prepending a mixin. Hence your problem.

I agree with @solnic. POLS violation here, better to write an explicit method for it. I see general consensus with @solnic comment, so I'm closing this.

**Comment by beauby** [comment]
@jodosha I don't understand your sentence "POLS violation here, better to write an explicit method for it.". Do you mean "The behavior of the framework is violating the principle of least surprise, so use a different name."? I would see how using a different name would be a temporary fix, but expecting people never to override inherited methods does not seem appropriate.

**Comment by jodosha** [comment]
@beauby Because `#create` is a public method exposed by default by a repository. Its role is to persist a single object into the database.

If we override `#create` to enhance it, this situation may "surprise" other devs in a project.

**Comment by beauby** [comment]
@jodosha I understand what you mean, and I agree that overriding `#create` is probably a bad practice. However, exposing a method that is **not overridable** is the actual surprising thing here, especially as one of the stated goals of hanami is to be less magical than rails and make use of POROs.

**Comment by jodosha** [comment]
@beauby I agree on the expectations about `#create`.

If you can, would you like to solve this problem? Thanks.

**Comment by beauby** [comment]
@jodosha Sure, I'll look into it – just wanted to make sure you were on board.

**Comment by jc00ke** [comment]
I've recently run into this issue and I think that @beauby is correct, in that in an OO language, *not* being able to override a public method violates `POLS`.

I'm working on an application where I need to hash and encrypt attributes in several models. Currently, I have to write a `CreateX` and `UpdateX` interactor for each model `X` instead of overriding `XRepository#create` and `XRepository#update`. In fact, having `XRepository#create` exposed "just because" means that I now have 2 ways to create an `X`, and in my case that's actually dangerous. Under no circumstances should someone be able to `#create` w/o hashing and encrypting.

@jodosha is there a technical reason because we both subclass and prepend? Is there another place we can talk about supporting this? Thanks!

**Comment by smarthouseprojectdotorg** [comment]
I agree with @jc00ke and @beauby and using the recent version of Hanami still get the above issues, hence found this discussion. It's 2020 and I see the issue is closed but not solved. I'm adding a cache to a repository and was surprised by the fact that I could not override #update and #delete and call super and update the cache in the overridden methods. Why? @jodosha Could you suggest a nice solution for this issue? I'm doing cached_update and cached_delete now but this is really not good and hope it's a temporary solution.

**Comment by solnic** [comment]
@smarthouseprojectdotorg this will be fixed in hanami-model 2.0.0
